### ISSUE TITLE
Config object passed on login request is mutated to include extra scopes

### ISSUE THREAD

**Issue by mmuarc** [AzureAD/microsoft-authentication-library-for-js#846]
## I'm submitting a...

[ ] Regression (a behavior that used to work and stopped working in a new release)
[x ] Bug report  
[x ] Performance issue
[x ] Feature request
[ ] Documentation issue or request
[ ] Other... Please describe:
</code></pre>

## Browser:
- [x ] Chrome version XX
- [ ] Firefox version XX
- [ ] IE version XX
- [ ] Edge version XX
- [ ] Safari version XX

## Library version

Library version: X.0.2

## Current behavior
config object passed when login is mutated to include two extra scopes: 'openid', 'profile'

## Expected behavior
msal shouldn't mutate the passed config object

## Minimal reproduction of the problem with instructions
I had my configuration for the login request stored in a const var with all the scopes like this:

export const config = {
  auth: {
    clientId: process.env.REACT_APP_CLIENT_ID,
    redirectUri: process.env.REACT_APP_REDIRECT_CALLBACK,
    validateAuthority: false,
    authority: process.env.REACT_APP_AUTHORITY,
    b2cScopes: process.env.REACT_APP_SCOPES.split(','),
  },

I used that object to login and also the relevant scopes every time I had to get a token async. The issue is that after login, at some moment that I can't exactly determine the property b2cScopes is mutated to include two scopes that are not controlled by me: 'openid' and 'profile'.

In my acquireTokenSilent requests I am passing config.auth.b2cScopes for the requested scopes. Since that property contained these two extra scopes the caching mechanism failes to retrieve the cached token and reaches the server for  a new one.

To workaround the problem i just make sure that I pass explicitly the two scopes that I need instead of reusing my config object. But I don't think UserAgentApplication should mutate a passed in parameter.
 
What do you think?

Thanks,

Miguel

**Comment by pkanher617** [comment]
@mmuarc Thanks for a detailed description here.

I just wanted to clear up some usage with MSAL so I can make sure I understand your problem. We have two separate "configuration" objects in MSAL.js now. The first is the configuration of MSAL.js, which is described in Configration.ts, the second is the request object sent to the login and acquireToken APIs, which is described in AuthenticationParameters.ts. 

So to rephrase your issue: You already have an accessToken in cache for the b2cScopes that you have stored. You are sending these same scopes in the AuthenticationParameters request object to acquireTokenSilent to get this cached token, and the acquireTokenSilent request is bypassing the cache to retrieve a new token from the server. (Please let me know if this is incorrect).

We have been getting a few issues related to the openid and profile scopes. IF this is your issue, I'm a little surprised that this behavior is happening, since MSAL does not check for scopes with openid and profile. As long as **all** scopes that you requested (not including openid and profile) are present in the accessToken, we will perform a successful cache lookup. We also do not perform cache lookup with openid and profile scopes - only the scopes that you have given us. "openid" and "profile" are only added to the request urls when sending the login or acquireToken requests, which is a required behavior by the B2C auth service.

Are your B2C scopes hard-coded? If not, how are they generated for each request?

When you make a request, please check the scopes parameter of the response. These are the scopes sent back from the server, and the scopes that we use to cache the token. If any of your requested scopes are **not** present in the set of scopes in the response object, then those scopes have been rejected. If this behavior is happening, please let us know.

**Comment by mmuarc** [comment]
Hi @pkanher617, thanks for your response. I think I maybe didn't explain the issue correctly.

Consider a const object in my application, that I use to store the initial configuration for UserAgentApplication and also contains the scopes that I'm going to use during the whole application lifetime. The object is this:

```export const config= {
    auth: {
      clientId: process.env.REACT_APP_CLIENT_ID,
      redirectUri: process.env.REACT_APP_REDIRECT_CALLBACK,
      validateAuthority: false,
      authority: process.env.REACT_APP_AUTHORITY,
      b2cScopes: process.env.REACT_APP_SCOPES.split(','),
      // Updated
      extraQueryParameters: { "ui_locales": "xxx", "loginContext": "xxx" }
    },
    cache: {
      cacheLocation: 'localStorage',
      storeAuthStateInCookie: true
    }
}```

now, this same object I use as a source of truth for any interaction I have with msal. So for intialization I do this:
`const clientApplication = new UserAgentApplication(getConfig);`

for login I do this:


```
  static loginRedirect = () => {
    let loginRequest = {
      scopes: **config.auth.b2cScopes**,
      authority: config.auth.authority,
      extraQueryParameters: config.auth.extraQueryParameters
    };
    clientApplication
      .loginRedirect(loginRequest)
  };

```

and for each token I do this:

```
 static getTokenAsync = async () => {
    let tokenRequest = {
      **scopes: config.auth.b2cScopes,**
      authority: config.auth.authority,
      extraQueryParameters: config.auth.extraQueryParameters
    };
    return new Promise((resolve, reject) => {
      clientApplication.acquireTokenSilent(tokenRequest).then(function (loginResponse) {
        return resolve(loginResponse.accessToken);
      }).catch(function (error) {
        console.log(error);
        Ab2c.logout()
        let newError = new Error(error);
        newError.fromMsal = true;
        return reject(newError);
      });
    });
  }
```
As you can see, I'm actually using the same array for my scopes all the time, since it comes from the object initialized only once. So what happens is that after the login, MSAL includes openid and profile to the array b2cScopes. This is further used to get each token with the new scopes already included. And due to these extra two scopes being present in the request, the caching mechanism is not retrieving the token.

Here is a sloppy screenshot where you can see the caching mechanism about to fail, with the request  including the 'openid' and 'profile' scopes (added by msal during login) and the cached mechanism properly containing only the two relevant scopes.

![image](https://user-images.githubusercontent.com/39746984/61699894-aaae5580-ad33-11e9-83a1-46f7c9ead297.png)

**Comment by pkanher617** [comment]
@mmuarc I apologize, I don't think I fully understood your issue the first time. However, I think we identified where this is happening. Thanks for letting us know. I will try to push a fix for this.

**Comment by onlyann** [comment]
@pkanher617 If that helps, the `scopes` array passed to `acquireTokenSilent` gets mutated.
```Typescript
userAgentApp
      .acquireTokenSilent({
        scopes // scopes is mutated by msal
      }
```

**Comment by DarylThayil** [comment]
@mmuarc  this should be resolved in 1.1.2 (#875) please try this version and create a new issue if this is not resolved for you
